From a4b9092f39e57116297123041a4ae158773b64a1 Mon Sep 17 00:00:00 2001
From: jinghua <jinghua@marvell.com>
Date: Tue, 17 Nov 2015 20:14:11 -0800
Subject: [PATCH 0512/1240] fix: SPI: armada3700: add the missing GPIO
 configuration for SPI

	In A3700 Register Spec, it says that In North bridge GPIO configuration,
	bit 18 is for SPI quad mode, but this is not accurate description.
	In fact, bit 18 controls HOLD and WP pins for SPI, which is needed
	for all SPI modes, single, dual, and quad.

Change-Id: Ie106b65a2ebefacb2b2055e43b083f59c34442d2
Signed-off-by: jinghua <jinghua@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/24968
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
---
 arch/arm/include/asm/arch-armadalp/regs-base.h | 1 +
 drivers/misc/mvebu_gpio.c                      | 9 ++++++++-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/arch/arm/include/asm/arch-armadalp/regs-base.h b/arch/arm/include/asm/arch-armadalp/regs-base.h
index b76cc67..6050037 100644
--- a/arch/arm/include/asm/arch-armadalp/regs-base.h
+++ b/arch/arm/include/asm/arch-armadalp/regs-base.h
@@ -49,6 +49,7 @@
 
 #define MVEBU_A3700_GPIO_NB_SEL (MVEBU_REGS_BASE + 0x13830)
 #define MVEBU_A3700_GPIO_TW1_GPIO_EN_OFF (10)
+#define MVEBU_A3700_GPIO_SPI_GPIO_EN_OFF (18)
 
 /* clock registers */
 #define MVEBU_NORTH_CLOCK_REGS_BASE	(MVEBU_REGS_BASE + 0x13000)
diff --git a/drivers/misc/mvebu_gpio.c b/drivers/misc/mvebu_gpio.c
index 9dbef60..182e9d4 100644
--- a/drivers/misc/mvebu_gpio.c
+++ b/drivers/misc/mvebu_gpio.c
@@ -42,11 +42,18 @@ void mvebu_a3700_gpio(void)
 	writel(reg_val, MVEBU_A3700_GPIO_SB_SEL);
 
 	/*
-	  * I2C GPIO
+	  * I2C, SPI GPIO
 	  */
 	reg_val = readl(MVEBU_A3700_GPIO_NB_SEL);
 	/* enable GPIO for I2C */
 	reg_val = reg_val & (~(1 << MVEBU_A3700_GPIO_TW1_GPIO_EN_OFF));
+	/* enable GPIO for SPI
+	  * In A3700 Register Spec, it says that In North bridge GPIO configuration,
+	  * bit 18 is for SPI quad mode, but this is not accurate description.
+	  * In fact, bit 18 controls HOLD and WP pins for SPI, which is needed for all
+	  * SPI mode, single, dual, and quad.
+	*/
+	reg_val = reg_val & (~(1 << MVEBU_A3700_GPIO_SPI_GPIO_EN_OFF));
 	writel(reg_val, MVEBU_A3700_GPIO_NB_SEL);
 
 	return;
-- 
1.9.1

